File: Java/ToyORB/src/ToyORB/network/MessageServer.java
package ToyORB.network;

import ToyORB.messages.Message;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Map;

public class MessageServer {
	private ServerSocket serverSocket;

	public MessageServer(String serviceName, int port) throws IOException {
		Message.initReaders();
		serverSocket = new ServerSocket(port);
		System.out.println(serviceName + " server listening on port " + port);
	}

	public void processNextRequest(Map<Message.MessageType, MessageProcessor> messageProcessors) throws IOException {
		Socket socket = serverSocket.accept();

		DataInputStream recvStream = new DataInputStream(socket.getInputStream());
		Message request = Message.readFrom(recvStream);
		socket.shutdownInput();
		System.out.println("Read request from socket " + request);

		MessageProcessor messageProcessor = messageProcessors.get(request.type);
		if (messageProcessor == null) {
			System.out.println("Dropped request of unhandled type " + request.type);
			socket.close();
			return;
		}

		@SuppressWarnings("unchecked")
		Message response = messageProcessor.processRequest(request, socket.getInetAddress());

		System.out.println("Writing response to socket - " + response);
		DataOutputStream sendStream = new DataOutputStream(socket.getOutputStream());
		response.writeInto(sendStream);
		sendStream.flush();
		socket.shutdownOutput();

		recvStream.close();
		sendStream.close();
	}

	protected void finalize() throws Throwable {
		super.finalize();
		serverSocket.close();
	}
}

File: .NET Core/ToyORB/Network/MessageServer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Net.Sockets;
using ToyORB.Messages;

namespace ToyORB.Network
{
    public class MessageServer
    {
        private readonly TcpListener _serverListener;

        public MessageServer(string serviceName, int port)
        {
            Message.InitReaders();
            _serverListener = new TcpListener(IPAddress.Any, port);
            _serverListener.Start();
            Console.WriteLine($"{serviceName} server listening on port {port}");
        }

        public void RegisterProcessor<TReq, TRes>(IMessageProcessor<TReq, TRes> processor) where TReq : Message where TRes : Message
        {
            _messageProcessors[processor.RequestType] = processor;
        }

        private readonly Dictionary<Message.MessageType, IMessageProcessor> _messageProcessors = new Dictionary<Message.MessageType, IMessageProcessor>();

        public void ProcessNextRequest()
        {
            using (TcpClient socket = _serverListener.AcceptTcpClientAsync().Result)
            {
                var recvStream = new BinaryReader(socket.GetStream());
                Message request = Message.ReadFrom(recvStream);
                Console.WriteLine($"Read request from socket {request}");

                IMessageProcessor processor = _messageProcessors[request.Type];
                Message response = processor.ProcessRequest(request, socket.Client.RemoteEndPoint);

                Console.WriteLine($"Writing response to socket {response}");
                var sendStream = new BinaryWriter(socket.GetStream());
                response.WriteInto(sendStream);
                sendStream.Flush();
            }
        }

    }
}
